File: backend/internal/migration/prompts.go
// Package migration — 4 类会话迁移辅助提示词模板（Go 版）。
//
// 与 opencode-plugin/src/prompts.ts 对齐，保证 Pocket 端预拼和 plugin 端 fallback 一致。
// 4 类：
//   - env_sync:      检查 git/依赖/工作目录
//   - task_resume:   从上次 nextAction 续接
//   - result_verify: 验证上次产物
//   - acc_report:    阶段完成时向 ACC 上报
package migration

import (
	"fmt"
	"strings"

	"github.com/halfking/pocket-opencode/backend/internal/model"
)

// BuildPrompts 按选中模板拼接最终注入提示词。
// 顺序固定 env → resume → verify → report，保证环境检查永远在续接之前。
func BuildPrompts(brief *model.SessionResumeBrief, templates []string) string {
	if brief == nil {
		brief = &model.SessionResumeBrief{}
	}
	order := []string{"env_sync", "task_resume", "result_verify", "acc_report"}
	selected := make(map[string]bool)
	for _, t := range templates {
		selected[t] = true
	}

	title := brief.Title
	if title == "" {
		title = brief.SessionID
	}
	parts := []string{
		fmt.Sprintf("# 任务迁移续接\n来源会话：%s\n来源实例：%s", title, brief.InstanceID),
	}
	for _, t := range order {
		if !selected[t] {
			continue
		}
		switch t {
		case "env_sync":
			parts = append(parts, buildEnvSync(brief))
		case "task_resume":
			parts = append(parts, buildTaskResume(brief))
		case "result_verify":
			parts = append(parts, buildResultVerify(brief))
		case "acc_report":
			parts = append(parts, buildAccReport(brief))
		}
	}
	return strings.Join(parts, "\n\n---\n\n")
}

func buildEnvSync(b *model.SessionResumeBrief) string {
	// 工作目录由迁移命令的 workingDirectory 字段单独传递给目标实例，此处给通用指引。
	return strings.Join([]string{
		"## 环境同步（请在继续任务前完成）",
		"1. 进入迁移命令指定的工作目录",
		"2. 运行 `git status` 与 `git log --oneline -5`，确认当前分支与上次提交",
		"3. 如来源实例有未推送的提交，先 `git pull --rebase` 同步远端",
		"4. 检查依赖：根据语言（go.mod / package.json / requirements.txt）执行安装",
		"5. 如工作目录与本机路径不一致，做相应重映射后再继续",
	}, "\n")
}

func buildTaskResume(b *model.SessionResumeBrief) string {
	lines := []string{"## 任务续接（从上次进度继续，不要重头开始）"}
	if b.LastObjective != "" {
		lines = append(lines, fmt.Sprintf("- 原始目标：%s", b.LastObjective))
	}
	if b.CurrentState != "" {
		lines = append(lines, fmt.Sprintf("- 当前状态：%s", b.CurrentState))
	}
	if len(b.Decisions) > 0 {
		lines = append(lines, "- 已确定的决策：")
		for _, d := range b.Decisions {
			lines = append(lines, fmt.Sprintf("  • %s", d))
		}
	}
	if len(b.Blockers) > 0 {
		lines = append(lines, "- 上次阻塞：")
		for _, blk := range b.Blockers {
			lines = append(lines, fmt.Sprintf("  • %s", blk))
		}
	}
	if b.NextAction != "" {
		lines = append(lines, fmt.Sprintf("- 下一步（请直接从这里接续）：%s", b.NextAction))
	} else if b.CurrentState != "" {
		lines = append(lines, "- 请根据当前状态判断下一步并继续。")
	}
	return strings.Join(lines, "\n")
}

func buildResultVerify(b *model.SessionResumeBrief) string {
	if len(b.ChangedFiles) == 0 && len(b.Attachments) == 0 {
		return "## 成果验证\n（迁移包未提供已改文件清单，跳过验证步骤。）"
	}
	lines := []string{"## 成果验证（继续前先确认上次产物）"}
	if len(b.ChangedFiles) > 0 {
		lines = append(lines, "上次已修改/新增的文件，请逐一确认存在且内容完整：")
		for _, f := range b.ChangedFiles {
			lines = append(lines, fmt.Sprintf("  • `%s`", f))
		}
	}
	fileAtts := filterFileAtts(b.Attachments)
	if len(fileAtts) > 0 {
		lines = append(lines, "产物文件（如缺失请从链接重新拉取）：")
		for _, a := range fileAtts {
			lines = append(lines, fmt.Sprintf("  • %s — %s", a.Name, fallback(a.CloudReveURL, "(无URL)")))
		}
	}
	lines = append(lines, "确认后运行测试套件，全绿再继续；若缺失，先补齐再继续。")
	return strings.Join(lines, "\n")
}

func buildAccReport(b *model.SessionResumeBrief) string {
	return strings.Join([]string{
		"## ACC 汇报（阶段完成时执行）",
		"当本阶段任务完成或到达自然检查点时：",
		"1. 调用 `acc_task_complete` 标记任务完成（附简短成果说明）",
		"2. 若属于 Mission，调用 `acc_mission_report` 上报阶段进度",
		"3. 如遇阻塞需人类介入，调用 `acc_request_human_input`",
		fmt.Sprintf("来源任务标识（如适用）：%s", fallback(b.SessionID, "(无)")),
	}, "\n")
}

func filterFileAtts(atts []model.AttachmentRef) []model.AttachmentRef {
	var out []model.AttachmentRef
	for _, a := range atts {
		if a.Type == "file" || a.Type == "diff" {
			out = append(out, a)
		}
	}
	return out
}

func fallback(v, def string) string {
	if v == "" {
		return def
	}
	return v
}

File: opencode-plugin/src/prompts.ts
/**
 * 会话迁移辅助提示词模板（Phase 3）
 *
 * 4 类模板，按迁移包（SessionResumeBrief + messages + summary）拼接成一段
 * 注入到新会话首条 prompt 的文本，引导新实例：
 *   - env_sync:      检查 git/依赖/工作目录，确保环境一致
 *   - task_resume:   从上次 nextAction 续接，不重头开始
 *   - result_verify: 先验证上次产物（文件存在 + 测试通过），再继续
 *   - acc_report:    阶段完成时向 ACC 上报，驱动后续编排
 *
 * 同一份逻辑在 Pocket 的 internal/migration/prompts 包里有 Go 镜像实现，
 * 用于 Pocket 端预览/编辑提示词后再下发命令。
 */

/** 迁移包的最小结构（与 model.SessionResumeBrief + 附件对齐）。 */
export interface MigrationPack {
  session_meta?: {
    id?: string
    title?: string
    directory?: string
    instance?: string
  }
  resume_brief?: {
    current_state?: string
    last_objective?: string
    decisions?: string[]
    changed_files?: string[]
    blockers?: string[]
    next_action?: string
  }
  summary?: string
  attachments?: Array<{ name?: string; cloudreve_url?: string; type?: string }>
  messages?: Array<{ role?: string; content?: string }>
}

export type PromptTemplate = 'env_sync' | 'task_resume' | 'result_verify' | 'acc_report'

/**
 * 按选中的模板拼接最终的注入提示词。
 * 选中的模板按固定顺序（env → resume → verify → report）拼合，
 * 保证环境检查永远在续接之前。
 */
export function buildMigrationPrompts(
  pack: MigrationPack,
  templates: PromptTemplate[] = ['env_sync', 'task_resume', 'result_verify'],
): string {
  const order: PromptTemplate[] = ['env_sync', 'task_resume', 'result_verify', 'acc_report']
  const selected = order.filter((t) => templates.includes(t))

  const parts: string[] = []
  const title = pack.session_meta?.title || pack.session_meta?.id || '(未知会话)'
  parts.push(`# 任务迁移续接\n来源会话：${title}\n来源实例：${pack.session_meta?.instance || '(未知)'}`)

  for (const t of selected) {
    const block = TEMPLATE_BUILDERS[t](pack)
    if (block) parts.push(block)
  }

  return parts.join('\n\n---\n\n')
}

const TEMPLATE_BUILDERS: Record<PromptTemplate, (p: MigrationPack) => string> = {
  env_sync: (p) => {
    const dir = p.session_meta?.directory || '当前工作目录'
    return [
      '## 环境同步（请在继续任务前完成）',
      `1. 进入工作目录：\`${dir}\``,
      '2. 运行 `git status` 与 `git log --oneline -5`，确认当前分支与上次提交',
      '3. 如来源实例有未推送的提交，先 `git pull --rebase` 同步远端',
      '4. 检查依赖：根据语言（go.mod / package.json / requirements.txt）执行安装',
      '5. 如工作目录与本机路径不一致，做相应重映射后再继续',
    ].join('\n')
  },

  task_resume: (p) => {
    const b = p.resume_brief || {}
    const lines: string[] = ['## 任务续接（从上次进度继续，不要重头开始）']
    if (b.last_objective) lines.push(`- 原始目标：${b.last_objective}`)
    if (b.current_state) lines.push(`- 当前状态：${b.current_state}`)
    if (b.decisions?.length) {
      lines.push('- 已确定的决策：')
      for (const d of b.decisions) lines.push(`  • ${d}`)
    }
    if (b.blockers?.length) {
      lines.push('- 上次阻塞：')
      for (const blk of b.blockers) lines.push(`  • ${blk}`)
    }
    if (b.next_action) {
      lines.push(`- 下一步（请直接从这里接续）：${b.next_action}`)
    } else if (p.summary) {
      lines.push(`- 上次摘要：${p.summary}`)
      lines.push('- 请根据摘要判断下一步并继续。')
    }
    return lines.join('\n')
  },

  result_verify: (p) => {
    const files = p.resume_brief?.changed_files || []
    const atts = (p.attachments || []).filter((a) => a.type === 'file' || a.type === 'diff')
    if (files.length === 0 && atts.length === 0) {
      return '## 成果验证\n（迁移包未提供已改文件清单，跳过验证步骤。）'
    }
    const lines: string[] = ['## 成果验证（继续前先确认上次产物）']
    if (files.length) {
      lines.push('上次已修改/新增的文件，请逐一确认存在且内容完整：')
      for (const f of files) lines.push(`  • \`${f}\``)
    }
    if (atts.length) {
      lines.push('产物文件（如缺失请从链接重新拉取）：')
      for (const a of atts) lines.push(`  • ${a.name || '(文件)'} — ${a.cloudreve_url || '(无URL)'}`)
    }
    lines.push('确认后运行测试套件，全绿再继续；若缺失，先补齐再继续。')
    return lines.join('\n')
  },

  acc_report: (p) => {
    return [
      '## ACC 汇报（阶段完成时执行）',
      '当本阶段任务完成或到达自然检查点时：',
      '1. 调用 `acc_task_complete` 标记任务完成（附简短成果说明）',
      '2. 若属于 Mission，调用 `acc_mission_report` 上报阶段进度',
      '3. 如遇阻塞需人类介入，调用 `acc_request_human_input`',
      `来源任务标识（如适用）：${p.session_meta?.id || '(无)'}`,
    ].join('\n')
  },
}
